### PR TITLE
Quotation marks in properties resolver

### PR DESCRIPTION
**Pull request by biryukovmaxim**
added unit tests
new type in enum PropertyValue and Literal
added special case for parsing string included escaped chars

### CODE REVIEW

**Review by Wiezzel — COMMENTED** [review]
Renaming `val` to `other` adds a lot of noise here and seems unrelated to the issue.

**Comment by biryukovmaxim** [comment]
@Wiezzel,
I see, agree that this is not related to the issue. but without this renaming it is difficult to read/analyze/navigate the code that implements the comparison. the difference between val and value is not obvious. can I save changes in a separate commit of the same pr? Or should do another pr?

**Comment by Wiezzel** [comment]
@biryukovmaxim Another PR would be the best option from reviewer's point of view but making another commit in this PR seems acceptable for me.

File: core/market/resolver/src/resolver/prop_parser.rs [hunk 1]
```diff
@@ -209,14 +213,39 @@ named!(
...
+                ),
+                String::new(),
+                |acc, fragment| acc + str::from_utf8(fragment).unwrap()
+            ),
+            |str| Literal::String(str)
+        ),
+        char!('"')
+    ))
+);
+
```
Comment (mfranciszkiewicz) [inline core/market/resolver/src/resolver/prop_parser.rs:248]: Maybe I'm missing something, but I don't see how `Literal::String` would be chosen over `Literal::Str` while parsing.

If we skip the unescaping part, the `Literal::String` type wouldn't be needed and parsing could be done via e.g.:

```rust
named!(
    string_literal<Literal>,
    ws!(do_parse!(
        val: delimited!(
            tag!("\""),
            escaped!(none_of!("\\\""), '\\', one_of!("\\\"")),
            tag!("\"")
        ) >> (Literal::Str(str::from_utf8(val).unwrap()))
    ))
);
```

Comment (biryukovmaxim) [reply core/market/resolver/src/resolver/prop_parser.rs:248]: string_literal used here:
https://github.com/golemfactory/yagna/pull/1758/files#diff-b14d1e21873636609b0a7924e3830facdd507a38ad770f5f318c053774ca74acR446

I didn't find the way how I can parse string including escaped chars with nom crate version 2.*
I think it's not the best option, but acceptable workaround.

Another solution is upgrade nom and use something like this: https://docs.rs/nom/5.1.2/nom/combinator/fn.all_consuming.html in str_literal func

Comment (stranger80) [reply core/market/resolver/src/resolver/prop_parser.rs:248]: I must admit I'd prefer going in the direction pointed by @mfranciszkiewicz above. 
I get the need of introducing `Literal::String()` but when it comes to parsing escaped quotes, use the simpler macro written by Marek, just "unescape" the result (by replacing \" with ").

**Review by stranger80 — CHANGES_REQUESTED** [review]
I must admit I'd prefer going in the direction pointed by @mfranciszkiewicz above. 
I get the need of introducing `Literal::String()` but when it comes to parsing escaped quotes, use the simpler macro written by Marek, just "unescape" the result (by replacing \" with ").

Also, add relevant unit test in core/market/resolver/tests/prop-parser.rs, eg:

```
#[test]
fn parse_prop_value_from_literal_string_with_escaped_quotes() {
    assert_eq!(
        parse_prop_value_literal("\"\\\"king\\\" \\\"node\\\"\""),
        Ok(Literal::String("\"king\" \"node\""))
    );
}
```

File: core/market/resolver/src/resolver/prop_parser.rs (hunk 1 above)
Comment (biryukovmaxim) [reply core/market/resolver/src/resolver/prop_parser.rs]: I have made the parser simpler. Workaround is removed. Unit test is added

**Review by stranger80 — APPROVED** [review]
Yep, looks good, cheers!

**Review by mfranciszkiewicz — APPROVED** [review]

**Pull request merged**